crates/telio-pq/src/proto.rs

```diff
@@ -203,20 +221,52 @@ async fn handshake(
     Ok(TunnelSock { tunn, sock })
 }
 
-pub fn parse_get_response(pkgbuf: &[u8]) -> super::Result<kyber768::Ciphertext> {
+pub fn parse_get_response(pkgbuf: &[u8], local_port: u16) -> io::Result<kyber768::Ciphertext> {
     let ip = Ipv4Packet::new(pkgbuf).ok_or(io::Error::new(
         io::ErrorKind::InvalidData,
         "Invalid PQ keys IP packet received",
     ))?;
 
+    validate_get_response_ip(&ip).map_err(|reason| {
+        io::Error::new(io::ErrorKind::InvalidData, format!("PQ keys IP: {reason}"))
+    })?;
+
     let udp = UdpPacket::new(ip.payload()).ok_or(io::Error::new(
         io::ErrorKind::InvalidData,
         "Invalid PQ keys UDP packet received",
     ))?;
 
+    validate_get_response_udp(&udp, local_port).map_err(|reason| {
+        io::Error::new(io::ErrorKind::InvalidData, format!("PQ keys UDP: {reason}"))
+    })?;
+
     parse_response_payload(udp.payload())
 }
 
+fn validate_get_response_ip(ip: &Ipv4Packet) -> Result<(), &'static str> {
+    if ip.get_source() != REMOTE_IP {
+        return Err("invalid src IP");
+    }
+    if ip.get_destination() != LOCAL_IP {
+        return Err("invalid dst IP");
+    }
+    if ip.get_next_level_protocol() != ip::IpNextHeaderProtocols::Udp {
+        return Err("invalid protocol");
+    }
+    Ok(())
+}
+
+fn validate_get_response_udp(udp: &UdpPacket, local_port: u16) -> Result<(), &'static str> {
+    if udp.get_source() != SERVICE_PORT {
+        return Err("PQ keys UDP: invalid src port");
+    }
+    if udp.get_destination() != local_port {
+        return Err("PQ keys UDP: invalid dst port");
+    }
+
```

Comment:
Could we additionally validate UDP checksum?

I believe it should be enough just to compare [ipv4_chacksum()](https://docs.rs/pnet/latest/pnet/packet/udp/fn.ipv4_checksum.html) against [get_checksum()](https://docs.rs/pnet/latest/pnet/packet/udp/struct.UdpPacket.html#method.get_checksum)

Reply:
Sure

Reply:
Done